src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1233,126 @@ private int SkipBlockComment(int i)
+                case 'r': return "\r";
+                case 't': return "\t";
+                case 'u': return ScanUnicodeEscapeSequence();
+                case 'v': return "\v";
+                default: return c.ToString();
+            }
+        }
+
+        private string ScanUnicodeEscapeSequence()
+        {
+            int escSeqStartIndex = _currentIndex - 2;
+            int maxNumberOfHexDigits = 6;
+            bool isTerminated = false;
+
+            char c = GetChar();
+            if (c != '{')
+            {
+                UngetChar();
+
+                IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                return s_unknownUnicodeChar;
+            }
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            int i;
+            for (i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = GetChar();
+
+                // Sequence has been terminated.
+                if (c == '}')
+                {
+                    if (i == 0)
+                    {
+                        // Sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                        return s_unknownUnicodeChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if (!c.IsHexDigit())
+                {
+                    UngetChar();
+
+                    ReportError(_currentIndex, () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+
+                sb.Append(c);
+            }
+
+            if (!isTerminated)
+            {
+                c = GetChar();
+                if (c != '}')
+                {
+                    UngetChar();
+                    ReportError(_currentIndex,
+                        i == maxNumberOfHexDigits && c.IsHexDigit()
+                            ? (Expression<Func<string>>)(() => ParserStrings.TooManyDigitsInUnicodeEscapeSequence)
+                            : () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue = int.Parse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo);
+            if (unicodeValue <= Char.MaxValue)
+            {
+                return ((char)unicodeValue).ToString();
+            }
+            else if (unicodeValue <= 0x10FFFF)
+            {
+                return Char.ConvertFromUtf32(unicodeValue);
```

Comment:
Sorry that my description above was too short. I didn't mention bad UX.

Surrogate character support is repaid from history. We first got UTF7, then the early versions of UTF8 and UTF16 with fixed length(!). Then standards extend Unicode ranges. In UTF8 we do not see this because of its variable length nature. In the UTf16 case,  it had become a variable length (!) by surrogate characteres. At that point, all the languages that supported UTF were forced to add support for these extensions.
In fact, we have two Unicode standards which, at a later time, have made a covergence. From now on, we were able to direct any Unicode character as 32-bit hex and languages add this.
If we want to support direct encoding of characters, we need to support all options with fixed length (\x(ASCII/ANSI), \u7, \u8, \u16, \u32), but it doesn't make sense if we have a generic option `` `u{}`` to direct any Unicode characters as _Unicode code point_. Adding support surrogate characters in high level looks as useless trick which expose UTF16 limitations and leads to bad UX.


We have bad UX with limiting to 6 digits too. UCS4 is 32-bit and users can expect and is in the right to specify leading zeros. Compare:
```
`u10FFFF
`u0010FFFF
```
The leading zeros don't make sense but it is allowed hex form and allowed UCS4 32-bit form in the Unicode standard. Sometimes it can be used as good formating. So I expect users will wonder why we disable this.